File: Tests/CSharp/Normalizers/RequiredPortNormalizer.fs
// The MIT License (MIT)
// 
// Copyright (c) 2014-2015, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace Normalization

open System
open System.Linq
open NUnit.Framework
open Microsoft.CodeAnalysis
open Microsoft.CodeAnalysis.CSharp.Syntax
open Microsoft.CodeAnalysis.Diagnostics
open SafetySharp.Modeling
open SafetySharp.Compiler.Normalization
open SafetySharp.CSharp.Roslyn.Syntax
open SafetySharp.CSharp.Roslyn.Symbols

[<TestFixture>]
module RequiredPortNormalizer =
    let normalize csharpCode = 
        TestCompilation.GetNormalizedClass (RequiredPortNormalizer ()) (sprintf "using System.Diagnostics; %s" csharpCode)

    [<Test>]
    let ``does not normalize extern method not declared within a component class`` () =
        normalize "class X { extern void M(); }" =? "class X { extern void M(); }"

    [<Test>]
    let ``does not normalize non-extern component method`` () =
        normalize "class X : Component { public void M() {}}" =? "class X : Component { public void M() {}}"

    [<Test>]
    let ``normalizes extern 'void -> void' method within a component`` () =
        let normalized = sprintf "\
            class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__();\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] %s void M() => this.__reqPortField0__();}"

        normalize "class X : Component { public extern void M(); }" =? normalized "public"
        normalize "class X : Component { internal extern void M(); }" =? normalized "internal"
        normalize "class X : Component { protected internal extern void M(); }" =? normalized "protected internal"
        normalize "class X : Component { protected extern void M(); }" =? normalized "protected"
        normalize "class X : Component { private extern void M(); }" =? normalized "private"

    [<Test>]
    let ``normalizes extern void returning method within a component`` () =
        normalize "class X : Component { public extern void M(int a); }" =? 
            "class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__(int a);\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] public void M(int a) => this.__reqPortField0__(a);}"

        normalize "class X : Component { internal extern void M(int a, decimal b); }" =? 
            "class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__(int a, decimal b);\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] internal void M(int a, decimal b) => this.__reqPortField0__(a, b);}"

        normalize "class X : Component { internal extern void M(int a, decimal b, bool c); }" =? 
            "class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__(int a, decimal b, bool c);\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] internal void M(int a, decimal b, bool c) => this.__reqPortField0__(a, b, c);}"

    [<Test>]
    let ``normalizes extern method within params, ref, and out parameters`` () =
        normalize "class X : Component { public extern void M(ref int a, out bool b); }" =? 
            "class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__(ref int a, out bool b);\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] public void M(ref int a, out bool b) => this.__reqPortField0__(ref a, out b);}"

        normalize "class X : Component { internal extern void M(params int[] a); }" =? 
            "class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__(params int[] a);\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] internal void M(params int[] a) => this.__reqPortField0__(a);}"

    [<Test>]
    let ``normalizes extern non-void returning method within a component`` () =
        normalize "class X : Component { public extern bool M(int a); }" =? 
            "class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate bool __ReqPortDelegate0__(int a);\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] public bool M(int a) => this.__reqPortField0__(a);}"

        normalize "class X : Component { internal extern int M(int a, decimal b); }" =? 
            "class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate int __ReqPortDelegate0__(int a, decimal b);\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] internal int M(int a, decimal b) => this.__reqPortField0__(a, b);}"

        normalize "class X : Component { internal extern bool M(int a, decimal b, bool c); }" =? 
            "class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate bool __ReqPortDelegate0__(int a, decimal b, bool c);\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] internal bool M(int a, decimal b, bool c) => this.__reqPortField0__(a, b, c);}"

    [<Test>]
    let ``normalizes multiple required ports`` () =
        normalize "class X : Component { public extern bool M(int a);\npublic extern void M(); }" =? 
            "class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate bool __ReqPortDelegate0__(int a);\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] public bool M(int a) => this.__reqPortField0__(a);\n\
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate1__();\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate1__ __reqPortField1__;\
            [SafetySharp.Modeling.RequiredAttribute()] public void M() => this.__reqPortField1__();}"

    [<Test>]
    let ``normalizes explictly implemented extern method within a component`` () =
        normalize "namespace A.B.C { interface I { void M(); }} class X : Component, A.B.C.I { extern void A.B.C.I.M(); }" =?
            "class X : Component, A.B.C.I { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__();\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] void A.B.C.I.M() => this.__reqPortField0__();}"

    [<Test>]
    let ``preserves attributes applied to extern method within a component`` () =
        let normalized = sprintf "\
            class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__();\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            %s [SafetySharp.Modeling.RequiredAttribute()] void M() => this.__reqPortField0__();}"

        normalize "class X : Component { [DebuggerHidden] extern void M(); }" =? normalized "[DebuggerHidden]"
        normalize "class X : Component { [DebuggerHidden, DebuggerNonUserCode] extern void M(); }" =? normalized "[DebuggerHidden, DebuggerNonUserCode]"
        normalize "class X : Component { [DebuggerHidden] [DebuggerNonUserCode] extern void M(); }" =? normalized "[DebuggerHidden] [DebuggerNonUserCode]"

    [<Test>]
    let ``preserves line numbers of following lines`` () =
        let actual = normalize "class X : Component { public \nextern\n void M(int a,\nint b,\nint c); \n\nint f; }" |> normalizeNewLines
        let expected = 
            "class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__(int a, int b, int c);\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] public void M(int a, int b, int c) => this.__reqPortField0__(a, b, c);\n\n\n\n\n\nint f; }"

        actual =? expected

    [<Test>]
    let ``normalizes required port of component class nested in other component class`` () =
        let syntaxTree = TestCompilation.GetNormalizedSyntaxTree (RequiredPortNormalizer()) "class Y : Component { class X : Component { public extern void M(); }}"
        syntaxTree.Descendants<ClassDeclarationSyntax>().Single(fun c -> c.Identifier.ValueText = "X").ToFullString () =?  
            "class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__();\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] public void M() => this.__reqPortField0__();}"

    [<Test>]
    let ``normalizes required port of component class nested in other non-component class`` () =
        let syntaxTree = TestCompilation.GetNormalizedSyntaxTree (RequiredPortNormalizer()) "class Y { class X : Component { public extern void M(); }}"
        syntaxTree.Descendants<ClassDeclarationSyntax>().Single(fun c -> c.Identifier.ValueText = "X").ToFullString () =?  
            "class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__();\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] public void M() => this.__reqPortField0__();}"

    [<Test>]
    let ``does not normalize Update method`` () =
        normalize "class X : Component { public override void Update() {} }" =? "class X : Component { public override void Update() {} }"

    [<Test>]
    let ``normalizes replaced Update method`` () =
        normalize "class X : Component { public new extern void Update(); }" =? 
            "class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__();\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] public new void Update() => this.__reqPortField0__();}"

    [<Test>]
    let ``does not get confused by comments`` () =
        normalize "class X : Component { //TODO\nextern void M(); }" =? 
            "class X : Component { //TODO\n\
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__();\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            [SafetySharp.Modeling.RequiredAttribute()] void M() => this.__reqPortField0__();}"

        normalize "class X : Component { ///<summary>X</summary>\nextern void M(); }" =? 
            "class X : Component { \
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__();\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            ///<summary>X</summary>\n[SafetySharp.Modeling.RequiredAttribute()] void M() => this.__reqPortField0__();}"

        normalize "class X : Component { void N() {}\n///<summary>\n///X\n///</summary>\n//TODO\npublic extern void M(); }" =? 
            "class X : Component { void N() {}\n\
            [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private delegate void __ReqPortDelegate0__();\
            [System.Diagnostics.DebuggerBrowsableAttribute(System.Diagnostics.DebuggerBrowsableState.Never)] [System.Runtime.CompilerServices.CompilerGeneratedAttribute()] private __ReqPortDelegate0__ __reqPortField0__;\
            ///<summary>\n///X\n///</summary>\n//TODO\n[SafetySharp.Modeling.RequiredAttribute()] public void M() => this.__reqPortField0__();}"

File: Source/Compiler/Normalization/RequiredPortNormalizer.cs
// The MIT License (MIT)
// 
// Copyright (c) 2014-2015, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace SafetySharp.Compiler.Normalization
{
	using System;
	using System.Diagnostics;
	using System.Linq;
	using System.Runtime.CompilerServices;
	using CSharp.Roslyn;
	using CSharp.Roslyn.Symbols;
	using CSharp.Roslyn.Syntax;
	using Microsoft.CodeAnalysis;
	using Microsoft.CodeAnalysis.CSharp;
	using Microsoft.CodeAnalysis.CSharp.Syntax;
	using Modeling;

	/// <summary>
	///     Replaces all extern method declarations within a component with a matching delegate type, a field of that
	///     delegate type, and a non-extern method invoking the delegate instance stored in the field.
	/// 
	///     For instance:
	///     <code>
	///    		public extern void MyMethod(int a, double b);
	///    		// becomes (on a single line with uniquely generated names):
	///  		[CompilerGenerated] public delegate void d(int a, double b);
	///  		[DebuggerBrowsable(DebuggerBrowsableState.Never), CompilerGenerated] d f;
	///    		[SafetySharp.Modeling.RequiredAttribute] public void MyMethod(int a, double b) { f(a, b); }
	///    		
	///    		private extern bool MyMethod(out int a);
	///    		// becomes (on a single line with uniquely generated names):
	///    		[CompilerGenerated] public delegate bool d(out int a);
	///  		[DebuggerBrowsable(DebuggerBrowsableState.Never), CompilerGenerated] private d f;
	///    		[SafetySharp.Modeling.RequiredAttribute] private bool MyMethod(out int a) { return f(out a); }
	///  
	/// 		private extern bool MyProperty { get; set; } // TODO!
	///    		// becomes (on a single line with uniquely generated names):
	///    		[CompilerGenerated] public delegate bool d1();
	///   		[CompilerGenerated] public delegate void d2(bool value);
	///  		[DebuggerBrowsable(DebuggerBrowsableState.Never), CompilerGenerated] private d1 f1;
	///  		[DebuggerBrowsable(DebuggerBrowsableState.Never), CompilerGenerated] private d2 f2;
	///    		private bool MyProperty { [SafetySharp.Modeling.RequiredAttribute] get { return f1(); } [SafetySharp.Modeling.RequiredAttribute] set { f2(value); } }
	///   	</code>
	/// </summary>
	public class RequiredPortNormalizer : Normalizer
	{
		/// <summary>
		///     Represents the [CompilerGenerated] attribute syntax.
		/// </summary>
		private static readonly AttributeListSyntax CompilerGeneratedAttribute =
			SyntaxBuilder.Attribute(typeof(CompilerGeneratedAttribute).FullName).WithTrailingSpace();

		/// <summary>
		///     Represents the [Required] attribute syntax.
		/// </summary>
		private static readonly AttributeListSyntax RequiredAttribute = SyntaxBuilder.Attribute(typeof(RequiredAttribute).FullName);

		/// <summary>
		///     Represents the [DebuggerBrowsable(DebuggerBrowsableState.Never)] attribute syntax.
		/// </summary>
		private static readonly AttributeListSyntax BrowsableAttribute = SyntaxBuilder.Attribute(
			typeof(DebuggerBrowsableAttribute).FullName,
			SyntaxFactory.ParseExpression("System.Diagnostics.DebuggerBrowsableState.Never"));

		/// <summary>
		///     The number of required ports declared by the compilation.
		/// </summary>
		private int _portCount;

		/// <summary>
		///     Initializes a new instance.
		/// </summary>
		public RequiredPortNormalizer()
			: base(NormalizationScope.Components)
		{
		}

		/// <summary>
		///     Normalizes the <paramref name="classDeclaration" />.
		/// </summary>
		protected override ClassDeclarationSyntax NormalizeClassDeclaration(ClassDeclarationSyntax classDeclaration)
		{
			var originalMembers = classDeclaration.Members;
			var members = originalMembers;

			var i = 0;
			foreach (var member in originalMembers)
			{
				var method = member as MethodDeclarationSyntax;
				if (method != null && method.Modifiers.Any(SyntaxKind.ExternKeyword))
					NormalizeMethod(method, ref members, ref i);
				++i;
			}

			return classDeclaration.WithMembers(members);
		}

		/// <summary>
		///     Normalizes the given method declaration and adds the generated members to the member list at the given index.
		/// </summary>
		/// <param name="methodDeclaration">The method declaration that should be normalized.</param>
		/// <param name="members">The members of the containing type that should be updated.</param>
		/// <param name="index">The index where the generated members should be inserted.</param>
		private void NormalizeMethod(MethodDeclarationSyntax methodDeclaration,
									 ref SyntaxList<MemberDeclarationSyntax> members,
									 ref int index)
		{
			// Create the delegate
			var methodSymbol = methodDeclaration.GetMethodSymbol(SemanticModel);
			var methodDelegate = methodSymbol.GetSynthesizedDelegateDeclaration("ReqPortDelegate" + _portCount);
			methodDelegate = methodDelegate.AddAttributeLists(CompilerGeneratedAttribute);

			// Create the field
			var fieldName = IdentifierNameSynthesizer.ToSynthesizedName("reqPortField" + _portCount++);
			var field = SyntaxBuilder.Field(fieldName, methodDelegate.Identifier.ValueText, Visibility.Private, BrowsableAttribute);
			field = field.AddAttributeLists(CompilerGeneratedAttribute).AsSingleLine();

			// Add the [Required] attribute if it is not already present
			var originalDeclaration = methodDeclaration;
			if (!methodDeclaration.HasAttribute<RequiredAttribute>(SemanticModel))
				methodDeclaration = methodDeclaration.WithAttributeLists(methodDeclaration.AttributeLists.Add(RequiredAttribute));

			// Remove the 'extern' keyword from the method
			var externIndex = methodDeclaration.Modifiers.IndexOf(SyntaxKind.ExternKeyword);
			methodDeclaration = methodDeclaration.WithModifiers(methodDeclaration.Modifiers.RemoveAt(externIndex));

			// Invoke the delegate stored in the field
			var fieldReference = SyntaxFactory.ParseExpression("this." + fieldName);
			var arguments = methodDeclaration.ParameterList.Parameters.Select(parameter =>
			{
				var argument = SyntaxFactory.Argument(SyntaxFactory.IdentifierName(parameter.Identifier));

				if (parameter.Modifiers.IndexOf(SyntaxKind.RefKeyword) != -1)
					return argument.WithRefOrOutKeyword(SyntaxFactory.Token(SyntaxKind.RefKeyword));

				if (parameter.Modifiers.IndexOf(SyntaxKind.OutKeyword) != -1)
					return argument.WithRefOrOutKeyword(SyntaxFactory.Token(SyntaxKind.OutKeyword));

				return argument;
			});

			var argumentList = SyntaxFactory.SeparatedList(arguments);
			var body = SyntaxFactory.InvocationExpression(fieldReference, SyntaxFactory.ArgumentList(argumentList));
			var arrowExpression = SyntaxFactory.ArrowExpressionClause(body);
			methodDeclaration = methodDeclaration.WithExpressionBody(arrowExpression);
			methodDeclaration = methodDeclaration.NormalizeWhitespace().AsSingleLine().EnsureSameLineCount(originalDeclaration);

			// Now add the delegate, field, and modified method to the members collection, 
			// removing the original method declaration
			members = members.RemoveAt(index);
			members = members.Insert(index, methodDelegate);
			members = members.Insert(++index, field);
			members = members.Insert(++index, methodDeclaration);
		}
	}
}